OSDOCS#4112[OPECO][4.14] Multi-arch support for Operator projects

michaelryanpeter opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.14
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: OSDOCS-4112
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview:
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

michaelryanpeter commented on operators/operator_sdk/osdk-multi-arch-support.adoc:
```diff
@@ -0,0 +1,27 @@
+* Build an image file for every platform you want your Operator to support.
+* Build a manifest list that specifies the platforms that your Operator supports.
+* Set your Operator's node affinity to support multiple architecture compute nodes.
+
```
@jaypoulz Do you think we need to further explain the difference between building an Operator that supports installation on multi-archs vs an Operator that supports multi-arch compute nodes?

jaypoulz replied:
I think it may be worth a callout to docs about multi-arch compute nodes. If I understand correctly, @kelbrown20 helped write those docs, so she might be able to help find a suitable description.

jaypoulz commented on operators/operator_sdk/osdk-multi-arch-support.adoc:
```diff
@@ -0,0 +1,27 @@
+
+toc::[]
+
+Operator projects that support multiple architectures and operating systems, or _platforms_, can run on more Kubernetes and {product-title} clusters than Operator projects that support only a single platform. Example architectures include `amd64`, `arm64`, `ppc64le`, and `s390x`. Example operating systems include Linux, IBM Z, and Windows.
```
While IBM's z/OS is an operating system, I don't think you can build a container for it. I would probably keep the OS list to Linux and Windows. We sometimes see MacOS for client binaries.

michaelryanpeter replied:
Addressed in https://github.com/openshift/openshift-docs/pull/64355/commits/cdcc6e4df48bcafea8897b9038d5d83e07a0dec7

michaelryanpeter replied:
Addressed in https://github.com/openshift/openshift-docs/pull/64355/commits/a2a3c6dae140f265de1729bb7b4022f56b0dcd07

jaypoulz commented on modules/osdk-multi-arch-node-affinity.adoc:
```diff
@@ -0,0 +1,121 @@
+[id="osdk-multi-arch-node-affinity_{context}"]
+= Safe scheduling an Operator's node affinity for multiple architecture compute nodes
+
+Node affinity is a set of rules used by the scheduler to define a pod's placement. Set node affinity requirements to ensure that your Operator's workloads only schedule to compute nodes with architectures that are compatible with the images in the pod. 
```
Here is probably where I would add a comment about how node affinity is important to set when your operator is expected to work in a cluster with compute nodes of multiple architectures. You can learn more about multi-arch compute clusters in *doc link*.

jaypoulz commented on modules/osdk-multi-arch-validate.adoc:
```diff
@@ -0,0 +1,31 @@
+[source,terminal]
+----
+$ operator-sdk bundle validate ./bundle \
+  --select-optional name=multiarch
```
I'll try to grab you a snippet from the tool for clean and messy output.

jaypoulz commented on modules/osdk-multi-arch-node-affinity.adoc:
```diff
@@ -0,0 +1,121 @@
++
+[source,terminal]
+----
+$ grep <regex_for_PodSpec_and_PodTemplateSpec>
```
Because object types in yaml don't actually declare their name, the only time you'd see PodSpec or PodTemplateSpec listed out explicitly like this would be when you interact with an API - like in the Go operator's dynamic workload example.

The dumb workaround I use is that I scan the code for the keyword `containers` since I know it is mandatory for this to be defined in both object types. I recognize this is not the cleanest strategy, but it's the best I've come up with.

michaelryanpeter replied:
WDYT? Too hand-wavy?

https://github.com/openshift/openshift-docs/pull/64355/commits/a96f690fb2e136cab7956ac52c692a69eb964ba2

michaelryanpeter replied:
Then maybe we highlight some YAML examples to help customers orient themselves?

Are there are any false positives for the `containers` field? If so, what happens if node affinity is set for an object that has the containers field but can't accept node affinity terms?

jaypoulz reviewed (commented):
OK so I've left comments on the overall skeleton and provided extra context where I could.

Overall, this is already looking very good. I didn't have a chance to run through it with an example operator, but I will try to get to that early next week.

jaypoulz commented on modules/osdk-multi-arch-node-affinity.adoc:
```diff
@@ -0,0 +1,121 @@
+
+[WARNING]
+====
+If you do not set node affinity requirements and a container is scheduled to a compute node with an incompatible architecture, the pod fails with an `exec format` error. When the pod restarts and fails again with the same error, an `ImagePullBackoff` event occurs.
```
The team explained the subtlety in the error messages today to me.
The `exec format` error:
    occurs when image is a single manifest image which is run on a node with incompatible architecture
    
The `image pull back-off` error:
    occurs when the image is manifestlisted but does not contain the image manifest matching the node architecture

aleskandro replied:
The possible events for workloads are:

- `CrashLoopBackOff`: the Entrypoint will fail to run by showing `exec format error` in the logs (the image is a standard manifest);
- `ImagePullBackOff`: the image is a manifest list, but there is no manifest for the node's architecture on which the Pod is placed or the wrong node affinity terms are set.

michaelryanpeter replied:
@jaypoulz and @aleskandro 
Thanks for the feedback. I addressed this in https://github.com/openshift/openshift-docs/pull/64355/commits/11fca674b5bcb691552fd3f3717485f684e9ceb9. PTAL.

michaelryanpeter replied:
Thanks for the feedback @jaypoulz and @aleskandro 

I addressed your feedback in https://github.com/openshift/openshift-docs/pull/64355/commits/ebbc2832c259310e7feb82c9c59e9c30b1e20ead

PTAL at the update for accuracy.

michaelryanpeter commented on modules/osdk-multi-arch-node-affinity.adoc:
```diff
@@ -0,0 +1,121 @@
+              - linux
+----
+
+. Go-based Operator projects that use dynamically created workloads, might embed of `PodSpec` and `PodTemplateSpec` objects in the Operator's logic . The following example shows how to update a `PodSpec` object by using the Go API:
```
Good points. Thank you. I think I need to refactor the information architecture a little bit. I also need to add some additional resources to provide background information on multi-arch compute nodes and node affinity.

michaelryanpeter commented:
@aleskandro and @jaypoulz I'm sorry for not responding to all of your feedback yet. I had some bugs come up that needed to be addressed. I have also needed to process some of the comments to adjust the information architecture of the content. I will reach out if I get blocked on anything. Thanks for your help!

jaypoulz commented on modules/osdk-multi-arch-building-manifest-list.adoc:
```diff
@@ -0,0 +1,38 @@
++
+[source,docker]
+----
+GOARCH=$TARGETARCH <1>
```
Maybe I'm verifying from a newer operator-sdk (v1.30.0), but my default Dockerfile looks like:
```
FROM golang:1.19 as builder
ARG TARGETOS
ARG TARGETARCH
...
RUN CGO_ENABLED=0 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o manager main.go
```

jaypoulz commented on modules/osdk-multi-arch-building-manifest-list.adoc:
```diff
@@ -0,0 +1,38 @@
++
+[source,terminal]
+----
+$ docker buildx build --push \
```
I'm a little confused with this  command.

If i ran the `$ make docker-buildx IMG=<image_registry>/<organization_name>/<repository_name>:<version_or_sha>` above, that built and pushed the manager image. So am I left to assume that this is for any non-manager images in my operator? Should that be said more explicitly?

jaypoulz replied:
Also, before you can run this command, you need to initialize the docker buildx driver:
```
 docker buildx create --use
 ```
 
 Lastly, we should probably note that the `.` in the last part of the command is the context path. (Basically where it's going to look for a file called Dockerfile and things referenced in it). You can specificy an alternative filename with the `--file` parameter. It may be worth adding that explicitly to make it obvious how to configure it e.g.:
 `--filename Dockerfile`

michaelryanpeter replied:
So, I was very confused about this part of the upstream and kubebuilder docs. 

Is there a chance that I am just duplicating the same steps in slightly different ways? 

If we can leverage kubebuilder and the SDK to do this part, I would love to cut the redundant module.

jaypoulz commented on modules/osdk-multi-arch-inspect-image-manifest.adoc:
```diff
@@ -0,0 +1,73 @@
+$ docker manifest inspect <image>
+----
+where:
+<image>:: Specifies the Operator's image manifest, such as `docker://ubi9:latest`.
```
I think the sample image is called:
`docker://redhat/ubi9:latest`

jaypoulz commented on modules/osdk-multi-arch-inspect-image-manifest.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,terminal]
+----
+$ skopeo inspect --raw <image>
```
Verified :heavy_check_mark:

jaypoulz commented on modules/osdk-multi-arch-inspect-image-manifest.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,terminal]
+----
+$ skopeo inspect <image>
```
Verified :heavy_check_mark:

jaypoulz replied:
Image name when using skopeo should be `docker://redhat/ubi9:latest`
In other words, if you're using skopeo, you need to specify the transport for docker images. This applies to the verified command above as well.

jaypoulz commented on modules/osdk-multi-arch-inspect-image-manifest.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,terminal]
+----
+$ docker manifest inspect <image>
```
Verified :heavy_check_mark:

jaypoulz commented on modules/osdk-multi-arch-inspect-image-manifest.adoc:
```diff
@@ -0,0 +1,73 @@
++
+[source,terminal]
+----
+$ docker inspect <image>
```
Verified :heavy_check_mark: 
<image> name when using docker for this command and manifest inspect should be `redhat/ubi9:latest`

jaypoulz reviewed (changes requested):
Review is still in progress. I will pick up where I left off starting with the changes in `osdk-multi-arch-node-affinity.adoc`

michaelryanpeter commented:
@jaypoulz and @everettraven I have reorganized the content based on both of your comments.

In particular, I simplified the procedure for making a manifest list, removing the steps where I unintentionally repeated myself. 

I also moved the validation procedure to the Bundle validate section of the docs. Lastly, I decided to link out to the procedure on adding labels to the Operator's CSV. That module needs to be rewritten to conform with this new content, and I am not going to be able to get to that before GA.

michaelryanpeter commented on modules/osdk-multi-arch-node-affinity.adoc:
```diff
@@ -0,0 +1,121 @@
+        - key: kubernetes.io/os
+            operator: In
+            values:
+              - linux
```
@aleskandro I created a ticket to track follow up work. I have included creating a module of example node affinity rules. You can track the work in https://issues.redhat.com/browse/OSDOCS-7997.

michaelryanpeter commented on operators/operator_sdk/osdk-multi-arch-support.adoc:
```diff
@@ -0,0 +1,47 @@
+== Next steps
+
+* xref:../../operators/operator_sdk/osdk-generating-csvs.adoc#olm-enabling-operator-for-multi-arch_osdk-generating-csvs[Label the platforms your Operator supports for Operator Lifecycle Manager (OLM)]
+* Bundle your Operator and Deploy with OLM
```
@everettraven Do you have to re-bundle your Operator after defining node affinity rules? Or will running the `make docker-buildx` command do the job?

everettraven replied:
You would have to re-bundle. The `make docker-buildx` builds the operator manager image itself, but not the manifests that would be used by OLM

aleskandro commented on modules/osdk-multi-arch-node-preference.adoc:
```diff
@@ -0,0 +1,63 @@
+      image: docker.io/<org>/<image_name>
+  affinity:
+      nodeAffinity:
+        preferredDuringSchedulingIgnoredDuringExecution: <1>
```
Hi @michaelryanpeter that's ok to move forward for another jira, but what we say here seems still a bit unsorted to me, but maybe I miss some context.

Why are we talking more about preferred affinity and missing the required affinity which is the critical one?

@jaypoulz @Prashanth684 any context I' missing here?


https://github.com/openshift/openshift-docs/pull/64355/files#diff-8c533024c02a4b9ed134ba9f297cef3fe2ab12369d09fe1f6dd8ba4661e36155R43-R57

aleskandro replied:
Ignore the thread... I misread.

aleskandro commented:
/label qe-approved

everettraven commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,119 @@
+[id="osdk-multi-arch-building-images_{context}"]
+= Building a manifest list of the platforms your Operator supports
+
+You can use the `make docker-buildx` command to build a manifest list of the platforms supported by your Operator and operands. A manifest list points to specific image manifests for one or more architectures. An image manifest specifies the platforms that an image supports.
```
Feel free to correct me if I'm misinterpreting the question. The way I'm interpreting this question is "Is the Operator-SDK opinionated on how operand images  (i.e images used in a Deployment created by the operator) are built?".

Short answer is no. I think a pretty common pattern for operators is deploying some form of an application, an example being the Operator-SDK tutorial of building an operator for deploying memcached. In the example, the creation of a `Memcached` CR results in a deployment being created that uses some other OCI image for memcached. The key here is that the Operator and the application are two separate things and it is assumed the application has it's own process for building images for the architectures supported (typically some variation of `docker buildx`).  The step in the prerequisites a bit further down is one that makes you aware of what architectures the application image itself supports and that helps you in deciding what architectures your operator should support, with the common case being to support the same architectures the operand image(s) support.

For an operator that doesn't have any operands that end up being workload type resources (think something like an operator that creates RBAC, secrets, configmaps, etc.) it is up to the operator authors to evaluate what architectures it is likely their user's nodes will be running and build for those architectures.

jaypoulz replied:
@michaelryanpeter I think the key thing here is - do the docs focus on explaining who to build and distribute an operator or both the operator and its operands?

It seems like a slippery slope to try to explain how different applications should be built, since its going to different per application. That said, we should should have a note somewhere that says something like, hey, if you're building the operator for multiple architectures, your application should probably support them too. You can find more support for building containers for multiple arches here {Insert-link-to-kubebuilder}.

michaelryanpeter replied:
I will work on that note. Unfortunately, we have rules from support about what kinds upstream documentation we are allowed to link to. Basically, we don't link to upstream procedures, since their support status is unclear. I will try to think through that a bit.

michaelryanpeter replied:
@everettraven and @jaypoulz: Do you think[ this admonition](https://github.com/openshift/openshift-docs/pull/64355/files#diff-4acb0b31c4e1bd9042a81b7a8fa87ebf61d412bfb9100d182acb7af42277a3b0R13-R16) is enough for now?

jaypoulz replied:
Alongside https://github.com/openshift/openshift-docs/pull/64355/files#diff-4acb0b31c4e1bd9042a81b7a8fa87ebf61d412bfb9100d182acb7af42277a3b0R33 I think this covers it. :)

everettraven replied:
+1, I think that is sufficient

michaelryanpeter commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,119 @@
+
+.Procedure
+
+. Inspect the image manifest of your Operator's base image to find which platforms your Operators can support. Run the following command to inspect your Operator's base image:
```
@everettraven 

Is it the same docker command to inspect workload images? 
```suggestion
. Inspect the image manifest of your Operator's base image to find which platforms your Operators can support. Run the following command to inspect your Operator's base image:

. If your Operator creates workload resources, inspect the workload images to find which architectures are supported:
[source,terminal]
----
docker manifest inspect <workload_image>
-----
```

jaypoulz replied:
The commands are the same, yes. :)

michaelryanpeter replied:
@everettraven and @jaypoulz : [Addressed here](https://github.com/openshift/openshift-docs/pull/64355/files#diff-4acb0b31c4e1bd9042a81b7a8fa87ebf61d412bfb9100d182acb7af42277a3b0R25-R33)

jaypoulz commented on modules/osdk-multi-arch-node-reqs.adoc:
```diff
@@ -0,0 +1,127 @@
+                                {
+                                    Key:      "kubernetes.io/arch",
+                                    Operator: "In",
+                                    Values:   []string{"amd64"},
```
Part of me wonders if it would be valuable to tweak this so that it specifies the same arches as:
https://github.com/openshift/openshift-docs/pull/64355/files#diff-fe05bf53a8124e691733d4aa0161cf19e4af47df547ab5fb37a45d5d8f929abaR61-R64

```
Values:   []string{"amd64","arm64","ppc64le","s390x"},
```

michaelryanpeter replied:
Updated: https://github.com/openshift/openshift-docs/pull/64355/commits/ecc4c26671ec320671fe037616d8a24cecee75e3#diff-fe05bf53a8124e691733d4aa0161cf19e4af47df547ab5fb37a45d5d8f929abaR92

jaypoulz commented on modules/osdk-multi-arch-node-reqs.adoc:
```diff
@@ -0,0 +1,127 @@
+* If you do not set node affinity rules and a container is scheduled to a compute machine with an incompatible architecture, the pod fails and triggers one of the following events:
++
+`CrashLoopBackOff`:: Occurs when an image manifest's entry point fails to run and an `exec format error` message is printed in the logs.
+`ImagePullBackOff`:: Occurs when a manifest list does not include a manifest for the architecture where a pod is scheduled or the node affinity terms are set to the wrong values.
```
Another case for when you set the node affinity terms incorrectly would be that your pod is unschedulable, but I don't think it warrants a mention since the message should clearly state why the workload doesn't match each node.

jaypoulz commented on modules/osdk-multi-arch-validate.adoc:
```diff
@@ -0,0 +1,31 @@
+.Example error message
+[source,text]
+----
+Some output
```
If your manager image is missing CSV labels:
```
ERRO[0016] Error: Value test-operator.v0.0.1: not all images specified are providing the support described via the CSV labels. Note that (SO.architecture): (linux.ppc64le) was not found for the image(s) [quay.io/jpoulin/test-operator:v1alpha1] 
ERRO[0016] Error: Value test-operator.v0.0.1: not all images specified are providing the support described via the CSV labels. Note that (SO.architecture): (linux.s390x) was not found for the image(s) [quay.io/jpoulin/test-operator:v1alpha1] 
ERRO[0016] Error: Value test-operator.v0.0.1: not all images specified are providing the support described via the CSV labels. Note that (SO.architecture): (linux.amd64) was not found for the image(s) [quay.io/jpoulin/test-operator:v1alpha1] 
ERRO[0016] Error: Value test-operator.v0.0.1: not all images specified are providing the support described via the CSV labels. Note that (SO.architecture): (linux.arm64) was not found for the image(s) [quay.io/jpoulin/test-operator:v1alpha1] 
```

If your manager image is healthy:
```
INFO[0020] All validation tests have completed successfully 
```

jaypoulz replied:
All of that said, it appears I forgot to bump the version of operator-framework/api used by operator-sdk. This means that the validate provided out of the box won't scan for mismatched affinity configurations. We may want to make that another follow-up task so that we can update the docs with the output that includes more information about affinity.

jaypoulz replied:
Here is the current output if your manager is successfully manifest-listed but you forgot the operatorhub flags:
```
WARN[0014] Warning: Value test-operator.v0.0.1: check if the CSV is missing the label (operatorframework.io/arch.<value>) for the Arch(s): ["amd64" "arm64" "ppc64le" "s390x"]. Be aware that your Operator manager image ["quay.io/jpoulin/test-operator:v1alpha1"] provides this support. Thus, it is very likely that you want to provide it and if you support more than amd64 architectures, you MUST,use the required labels for all which are supported.Otherwise, your solution cannot be listed on the cluster for these architectures 
```

michaelryanpeter replied:
Output examples added. @jaypoulz Do you think we need to add an admonition stating that the `multiarch` option does not validate for node affinity?

Or just fix it soon and update?

jaypoulz replied:
I think it's an enhancement moreso than an omission.
That said, I'll submit the PR today, and it will come in v1.32.0 of Operator SDK at the soonest.

jaypoulz replied:
So I would probably provide the user with the info we have released today, and then follow up with the updates in a Z-stream when the new version of Operator SDK is released.

michaelryanpeter replied:
Addressed in d0e842feaeba23ceeec245e52a74cba537e282b0

WDYT?

jaypoulz replied:
New commit (https://github.com/openshift/openshift-docs/commit/d0e842feaeba23ceeec245e52a74cba537e282b0) makes sense. A few minor notes:
1. You used the British spelling of labelled (IDK if that matters) :)
2. IDK if it's appropriate to mention that a future update to operator SDK will also include checks for node affinity. I think probably we leave that bit for the z-stream fixup because it will also touch up the success / error output sections.

michaelryanpeter replied:
We try not to talk about the future in docs. It leads to trouble more often than not. :)

And oops on the British spelling! Should be fixed now.

jaypoulz replied:
Excellent. :) The PR above is already merged, but it makes no sense to mention a release (v1.32.0) of operator SDK when it doesn't exist yet. I think all my comments/insights are resolved at this point.

jaypoulz reviewed (approved):
/lgtm

michaelryanpeter commented:
/label peer-review-needed

skopacz1 commented on modules/osdk-multi-arch-validate.adoc:
```diff
@@ -0,0 +1,43 @@
+You can validate your Operator's multi-platform readiness by running the `bundle validate` command. The command verifies that your Operator project meets the following conditions:
+
+* Your Operator's manager image supports the platforms labeled in the cluster service version (CSV) file.
+* Your Operator's CSV has labels for the supported platforms for Operator Lifecycle manager (OLM) and OperatorHub.
```
I think manager should be capitalized here:
```suggestion
* Your Operator's CSV has labels for the supported platforms for Operator Lifecycle Manager (OLM) and OperatorHub.
```

skopacz1 reviewed (commented):
Left some comments, but looks good to me!

skopacz1 commented:
/remove-label peer-review-needed
/remove-label peer-review-in-progress

/label peer-review-done

michaelryanpeter commented on modules/olm-enabling-operator-for-multi-arch.adoc:
```diff
@@ -1,6 +1,7 @@
 // Module included in the following assemblies:
 //
 // * operators/operator_sdk/osdk-generating-csvs.adoc
+// * operators/operator_sdk/osdk-multi-arch-support.adoc
```
Good call, that is an artifact I forgot to remove.

michaelryanpeter commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,125 @@
+----
+<1> Change the `GOARCH` field from `amd64` to `$TARGETARCH`.
+
+. Your Operator project's makefile defines the `PLATFORM` environment variable. When you run the `make docker buildx` command to generate a manifest list, the Operator SDK creates an image manifest for each of the platforms specified by the variable.
```
I will rework this a little. It is tricky because you need the context, but I also hate burying the lead.

michaelryanpeter commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,125 @@
+If you edit the `PLATFORMS` environment variable from the CLI, skip the next step in this procedure.
+====
+
+. Run the following command from your Operator project directory to build a manager image to cross-compile support for multiple platforms and push the manifest list your registry:
```
Good point. I will break it into smaller sentences

michaelryanpeter replied:
The command does a kitchen sink of functions, and the fact that it will build and push the image isn't obvious from the command syntax. I will finesse this a bit.

michaelryanpeter replied:
Please let me know if the updated wording is clearer.

skopacz1 replied:
In terms of sentence structure, it's a much much easier read after your update.

michaelryanpeter commented on modules/osdk-multi-arch-node-reqs.adoc:
```diff
@@ -0,0 +1,127 @@
+  affinity:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution: <1>
+	nodeSelectorTerms: <2>
+	- matchExpressions: <3>
+	  - key: kubernetes.io/arch <4>
+	    operator: In
+	    values:
+	    - amd64
+	    - arm64
+	    - ppc64le
+	    - s390x
+	  - key: kubernetes.io/os <5>
+	    operator: In
+	    values:
+		- linux
```
Ack. Unfortunately, my IDE has been switching from spaces to tabs randomly. I had it fixed earlier, but some customization must be reverting it. Good call.

michaelryanpeter commented on modules/osdk-multi-arch-node-reqs.adoc:
```diff
@@ -0,0 +1,127 @@
++
+where:
+
+`RequiredDuringSchedulingIgnoredDuringExecution`:: Defines a required rule.
+`NodeSelectorTerms`:: If you specify multiple `nodeSelectorTerms` associated with `nodeAffinity` types, then the pod can be scheduled onto a node if one of the `nodeSelectorTerms` is satisfied.
+`MatchExpressions`:: If you specify multiple `matchExpressions` associated with `nodeSelectorTerms`, then the pod can be scheduled onto a node only if all `matchExpressions` are satisfied.
+`kubernetes.io/arch`:: Specifies the architectures defined in the manifest list.
+`kubernetes.io/os`:: Specifies the operating systems defined in the manifest list.
```
This is actually a standard way to use a description list if code callouts are not the ideal way to markup a code example. It is described in the [repo guidelines.](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) (You have to scroll, or Crtl+f "where:"

skopacz1 replied:
The more you know. Thanks for teaching me something new!

michaelryanpeter commented on modules/osdk-multi-arch-node-reqs.adoc:
```diff
@@ -0,0 +1,127 @@
+
+* If you do not set node affinity rules and a container is scheduled to a compute machine with an incompatible architecture, the pod fails and triggers one of the following events:
++
+`CrashLoopBackOff`:: Occurs when an image manifest's entry point fails to run and an `exec format error` message is printed in the logs.
+`ImagePullBackOff`:: Occurs when a manifest list does not include a manifest for the architecture where a pod is scheduled or the node affinity terms are set to the wrong values.
```
It is a definition/description list. If you are creating a list of definitions, this is the asciidoc equivalent of HTML's [`<dl>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl).

michaelryanpeter replied:
But you are right, this isn't very actionable. I switched it to a warning admonition.

michaelryanpeter commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,125 @@
+...
+----
+
+** If the previous command does not output platform information, then the specified base image might be a single image instead of an image manifest. You can find which architectures an image supports by running the following command:
```
So, a single substep is not sequential, So, since there isn't a second step, it would be a bullet in this case. Just like a procedure with one step. That said, I have gone back and forth as to whether this should be inline or a substep. I will move inline for clarity.

michaelryanpeter commented on modules/osdk-multi-arch-building-images.adoc:
```diff
@@ -0,0 +1,125 @@
+# ...
+----
++
+You can edit the `PLATFORM` environment variable to specify the image manifests included in your manifest list. The following example defines the supported platforms as `linux/arm64` and `linux/amd64`:
```
This step is more conditional than optional (see the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-procedures#indicating-optional-and-conditional-steps). You were right that the original was confusing and poorly worded. Would you PTAL and let me know if this reads better?

skopacz1 replied:
Looks perfect, thanks for your update!

michaelryanpeter commented on modules/osdk-multi-arch-node-preference.adoc:
```diff
@@ -0,0 +1,63 @@
+
+.Prerequisites
+
+* An Operator project created or maintained with Operator SDK {osdk_ver} or later.
```
I took your advice on parallel formatting, but went a little slantwise with it :)

skopacz1 reviewed (commented):
A few more comments, let me know what you think!

michaelryanpeter commented:
/cherrypick enterprise-4.14
